Removal of the lang feature gate tests whitelist

PR #38914 has added a tidy check for gate tests in compile-fail for all unstable lang features, helping ensure that they are actually unstable. See also issue #22820.

To be recognized as gate test for a specific feature, tests either need to follow a naming scheme `feature-gate-NAME-OF-THE-FEAT.rs` or include a `// gate-test-NAME-OF-THE-FEAT` comment. For most of the currently present features, I could easily find matching tests in the test suite. Sadly, [there is still a whitelist](https://github.com/rust-lang/rust/blob/6fe23719feb20c25a8b2693d1f196e72d8da9af5/src/tools/tidy/src/features.rs#L168-L175) of currently 20 features for which it tolerates that it can't detect any gate tests.

The goal is to eliminate that whitelist, which this issue is tracking.

The features on the whitelist fall into the following categories:

**Has already tests, they just need to be marked (found this out only on a second look):**

* [x] `more_struct_aliases` → marked by PR #39243
* [x] `custom_attribute` → marked by PR #39243
* [x] `default_type_parameter_fallback` → marked by PR #39243
* [x] `use_extern_macros` → marked by PR #39243
* [x] `quote` → marked by PR #39250
* [x] `macro_reexport` → marked by PR #39250
* [x] `inclusive_range_syntax` → marked by PR #40031

**Used nowhere in the codebase. Remove?**

*got all removed by PR #39071*

* [x] `safe_suggestion`
* [x] `pushpop_unsafe`

**Needs a test (or I just haven't found a suiting test to mark, even on second look):**

* [x] `cfg_target_has_atomic` → got a test by #40150
* [x] `simd` → got a test by #40101
* [x] `abi_ptx`→ got a test by #40168
* [x] `const_indexing` → got a test by PR #39700
* [x] `stmt_expr_attributes` → got a test by PR #40252
* [ ] `unwind_attributes` → will get a test by PR #40279
* [x] `static_recursion` → got a test by PR #39834
* [ ] `cfg_target_thread_local` → will get a test by PR #40279
* [x] `unboxed_closures` → got a test by PR #40047
* [x] `staged_api` → got a test by PR #39772

**Don't know what to do with this one. Any ideas?**

* [x] `reflect` → removed by PR #39075

The "Needs a test" section is a good place for beginners to help out!

You basically need to:

0. if you haven't already, read `CONTRIBUTING.md` and `COMPILER_TESTS.md`
1. search in the `src/test/run-pass` suite for tests of that feature
2. copy one of those tests over to the compile-fail suite
3. rename it appropriately (`feature-gate-FEAT-NAME.rs`) and remove the `#![feature(...)]` line
4. remove the feature from the whitelist
5. test locally (`./x.py test src/tools/tidy` and `./x.py test src/test/compile-fail --test-args feature-gate-FEAT-NAME`)
5. submit a PR!

It would also be very good if you searched the `compile-fail` testsuite for already existing gate tests and mark those instead. A good place to look is the PR that introduced that feature, they usually also add gate tests.

You can help out in the "Has already tests" section as well. Here, contributing is even easier!

You'll need to:

0. read `CONTRIBUTING.md` and `COMPILER_TESTS.md` (second one is not strictly required, but it never harms :D)
1. add a line with `// gate-test-FEATURE_NAME` to the files that were mentioned above
2. remove the feature from the whitelist
3. test locally (`./x.py test src/tools/tidy`, since you only added comments)
4. submit a PR!

If you need help, ask in this thread. Also, its best to prevent duplicate work, so it would be nice if you could announce that you work on some of the items. Thanks in advance!

When I run the tidy check, everything run ok, but the following command produces an error: 

```
$ ./x.py test src/test/compile-fail --test-args feature-gate-more_struct_aliases

Finished debug [unoptimized] target(s) in 0.0 secs
Synchronizing submodule url for 'src/compiler-rt'
Synchronizing submodule url for 'src/jemalloc'
Synchronizing submodule url for 'src/liblibc'
Synchronizing submodule url for 'src/llvm'
Synchronizing submodule url for 'src/rt/hoedown'
Synchronizing submodule url for 'src/rust-installer'
HEAD is now at a8fc4c1 Merge pull request #28 from xen0n/preprocessor-firefighting
HEAD is now at e058ca6 Change how the default zone is found
HEAD is now at e49e9bb Auto merge of #478 - jackpot51:patch-1, r=alexcrichton
HEAD is now at ceb177e Merge pull request #60 from japaric/gh38406
HEAD is now at a3736a0 Merge pull request #6 from intelfx/patch-1
HEAD is now at 4f99485 Merge pull request #54 from brson/docdir
Building stage0 std artifacts (x86_64-apple-darwin -> x86_64-apple-darwin)
warning: /Library/Developer/CommandLineTools/usr/bin/ranlib: file: /Users/user/projects/rust/build/x86_64-apple-darwin/stage0-std/x86_64-apple-darwin/release/build/compiler_builtins-0511dd5445a3a209/out/libcompiler-rt.a(addtf3.o) has no symbols
warning: /Library/Developer/CommandLineTools/usr/bin/ranlib: file: /Users/user/projects/rust/build/x86_64-apple-darwin/stage0-std/x86_64-apple-darwin/release/build/compiler_builtins-0511dd5445a3a209/out/libcompiler-rt.a(divtf3.o) has no symbols
warning: /Library/Developer/CommandLineTools/usr/bin/ranlib: file: /Users/user/projects/rust/build/x86_64-apple-darwin/stage0-std/x86_64-apple-darwin/release/build/compiler_builtins-0511dd5445a3a209/out/libcompiler-rt.a(multf3.o) has no symbols
warning: /Library/Developer/CommandLineTools/usr/bin/ranlib: file: /Users/user/projects/rust/build/x86_64-apple-darwin/stage0-std/x86_64-apple-darwin/release/build/compiler_builtins-0511dd5445a3a209/out/libcompiler-rt.a(powitf2.o) has no symbols
warning: /Library/Developer/CommandLineTools/usr/bin/ranlib: file: /Users/user/projects/rust/build/x86_64-apple-darwin/stage0-std/x86_64-apple-darwin/release/build/compiler_builtins-0511dd5445a3a209/out/libcompiler-rt.a(subtf3.o) has no symbols
warning: /Library/Developer/CommandLineTools/usr/bin/ranlib: file: /Users/user/projects/rust/build/x86_64-apple-darwin/stage0-std/x86_64-apple-darwin/release/build/compiler_builtins-0511dd5445a3a209/out/libcompiler-rt.a(trampoline_setup.o) has no symbols
    Finished release [optimized] target(s) in 0.0 secs
Copying stage0 std from stage0 (x86_64-apple-darwin -> x86_64-apple-darwin / x86_64-apple-darwin)
Building stage0 test artifacts (x86_64-apple-darwin -> x86_64-apple-darwin)
    Finished release [optimized] target(s) in 0.0 secs
Copying stage0 test from stage0 (x86_64-apple-darwin -> x86_64-apple-darwin / x86_64-apple-darwin)
Building stage0 compiler artifacts (x86_64-apple-darwin -> x86_64-apple-darwin)
warning: ../rustllvm/RustWrapper.cpp:151:3: warning: default label in switch which covers all enumeration values [-Wcovered-switch-default]
warning:   default:
warning:   ^
warning: ../rustllvm/RustWrapper.cpp:1235:3: warning: default label in switch which covers all enumeration values [-Wcovered-switch-default]
warning:   default:
warning:   ^
warning: ../rustllvm/RustWrapper.cpp:1268:3: warning: default label in switch which covers all enumeration values [-Wcovered-switch-default]
warning:   default:
warning:   ^
warning: ../rustllvm/RustWrapper.cpp:1282:3: warning: default label in switch which covers all enumeration values [-Wcovered-switch-default]
warning:   default:
warning:   ^
warning: 4 warnings generated.
    Finished release [optimized] target(s) in 0.2 secs
Copying stage0 rustc from stage0 (x86_64-apple-darwin -> x86_64-apple-darwin / x86_64-apple-darwin)
Copying stage1 compiler (x86_64-apple-darwin)
thread 'main' panicked at 'failed to copy `/Users/user/projects/rust/build/x86_64-apple-darwin/stage0-sysroot/lib/rustlib/x86_64-apple-darwin/lib/libflate-89345601b107c0b9.dylib` to `/Users/user/projects/rust/build/x86_64-apple-darwin/stage1/lib/libflate-89345601b107c0b9.dylib`: Permission denied (os error 13)', src/bootstrap/util.rs:53
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

--- a/src/test/parse-fail/range_inclusive_gate.rs
+++ b/src/test/parse-fail/range_inclusive_gate.rs
@@ -8,6 +8,8 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
+// gate-test-inclusive_range_syntax
+
 // Make sure that #![feature(inclusive_range_syntax)] is required.
 
 // #![feature(inclusive_range_syntax, inclusive_range)]

--- a/src/tools/tidy/src/features.rs
+++ b/src/tools/tidy/src/features.rs
@@ -116,7 +116,8 @@ pub fn check(path: &Path, bad: &mut bool) {
     });
 
     super::walk_many(&[&path.join("test/compile-fail"),
-                       &path.join("test/compile-fail-fulldeps")],
+                       &path.join("test/compile-fail-fulldeps"),
+                       &path.join("test/parse-fail"),],
                      &mut |path| super::filter_dirs(path),
                      &mut |file| {
         let filename = file.file_name().unwrap().to_string_lossy();
@@ -169,8 +170,7 @@ pub fn check(path: &Path, bad: &mut bool) {
         "abi_ptx", "simd",
         "cfg_target_has_atomic",
         "unboxed_closures", "stmt_expr_attributes",
-        "cfg_target_thread_local", "unwind_attributes",
-        "inclusive_range_syntax"
+        "cfg_target_thread_local", "unwind_attributes"
     ];
 
     // Only check the number of lang features.

